src/ocf_seq_cutoff.c

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright(c) 2012-2018 Intel Corporation
+ * SPDX-License-Identifier: BSD-3-Clause-Clear
+ */
+
+#include "ocf_seq_cutoff.h"
+#include "ocf_cache_priv.h"
+#include "ocf_priv.h"
+
+#define SEQ_CUTOFF_FULL_MARGIN \
+                (OCF_TO_EVICTION_MIN + OCF_PENDING_EVICTION_LIMIT)
+
+static inline bool ocf_seq_cutoff_is_on(ocf_cache_t cache)
+{
+	if (!ocf_cache_is_device_attached(cache))
+		return false;
+
+	return (ocf_freelist_num_free(cache->freelist) <=
+			SEQ_CUTOFF_FULL_MARGIN);
+}
+
+static int ocf_seq_cutoff_stream_cmp(struct ocf_rb_node *n1,
+		struct ocf_rb_node *n2)
+{
+	struct ocf_seq_cutoff_stream *stream1 = container_of(n1,
+			struct ocf_seq_cutoff_stream, node);
+	struct ocf_seq_cutoff_stream *stream2 = container_of(n2,
+			struct ocf_seq_cutoff_stream, node);
+
+	if (stream1->rw < stream2->rw)
+		return -1;
+
+	if (stream1->last < stream2->last)
+		return -1;
+
+	if (stream1->last > stream2->last)
+		return 1;
+
+	return 0;
+}
+
+void ocf_core_seq_cutoff_init(ocf_core_t core)
+{
+	struct ocf_seq_cutoff_stream *stream;
+	int i;
+
+	ocf_core_log(core, log_info, "Seqential cutoff init\n");
+	env_rwlock_init(&core->seq_cutoff.lock);
+	ocf_rb_tree_init(&core->seq_cutoff.tree, ocf_seq_cutoff_stream_cmp);
+	INIT_LIST_HEAD(&core->seq_cutoff.lru);
+
+	for (i = 0; i < OCF_SEQ_CUTOFF_MAX_STREAMS; i++) {
+		stream = &core->seq_cutoff.streams[i];
+		stream->last = 0;
+		stream->bytes = 0;
+		stream->rw = 0;
+		ocf_rb_tree_insert(&core->seq_cutoff.tree, &stream->node);
+		list_add_tail(&stream->list, &core->seq_cutoff.lru);
+	}
+}
+
+bool ocf_core_seq_cutoff_check(ocf_core_t core, struct ocf_io *io)
+{
+	ocf_seq_cutoff_policy policy = ocf_core_get_seq_cutoff_policy(core);
+	uint32_t threshold = ocf_core_get_seq_cutoff_threshold(core);
+	ocf_cache_t cache = ocf_core_get_cache(core);
+	struct ocf_seq_cutoff_stream item = { .last = io->addr, .rw = io->dir };
+	struct ocf_seq_cutoff_stream *stream;
+	struct ocf_rb_node *node;
+	bool result = false;
+
+	switch (policy) {
+		case ocf_seq_cutoff_policy_always:
+			break;
+
+		case ocf_seq_cutoff_policy_full:
+			if (ocf_seq_cutoff_is_on(cache))
+				break;
+			return false;
+
+		case ocf_seq_cutoff_policy_never:
+			return false;
+		default:
+			ENV_WARN(true, "Invalid sequential cutoff policy!");
+			return false;
+	}
+
+	env_rwlock_read_lock(&core->seq_cutoff.lock);
+	node = ocf_rb_tree_find(&core->seq_cutoff.tree, &item.node);
+	stream = container_of(node, struct ocf_seq_cutoff_stream, node);
+	if (stream && stream->bytes + io->bytes >= threshold)
+		result = true;
```

Comment:
Return value of  `ocf_rb_tree_find` isn't checked. It can and will be `NULL` - promotion policy tests are segfaulting because of that.

Reply:
Done.